### PR TITLE
Karen transition - mappings to structs

### PR DESCRIPTION
**Pull request by kds0718**

### CODE REVIEW

File: contracts/StateMachine.sol [hunk 1]
```diff
@@ -1,141 +1,147 @@
...
+
+        while (outgoing.length > 0) {
+            bool stateChanged = false;
+            //consider each of the next states in turn
+            for (uint256 j = 0; j < outgoing.length; j++) {
+                //Get the state that you are now to consider
+                bytes32 nextState = outgoing[j];
+                bytes32 transitionId = getTransitionId(currentStateId, nextState);
+                // If this state's start condition is met, go to this state and continue
+                function(bytes32) internal returns(bool)[] startConds = transitions[transitionId].startConditions;
```
Comment (alicevhenshaw) [inline contracts/StateMachine.sol]: You can't actually define an array like this strangely - the solidity compiler can't deal with the word function being used inside a function yet.

Comment (alicevhenshaw) [reply contracts/StateMachine.sol]: Even though truffle accepts it. You sadly have to always just use transitions[transitionId].startConditions :(

File: contracts/StateMachine.sol [hunk 2]
```diff
@@ -1,141 +1,147 @@
...
-    }
-
-}
+pragma solidity 0.4.19;
+
+
+contract StateMachine {
+
+    //Struct to hold elements of transition
+    struct Transitions {
```
Comment (alicevhenshaw) [inline contracts/StateMachine.sol]: Each struct is just 1 transition - so I would personally call it 'Transition' not 'Transitions' as it just represents 1 Transition. (Maybe I'm being picky here aha)

File: contracts/StateMachine.sol [hunk 3]
```diff
@@ -1,141 +1,147 @@
...
+    }
+
+    //Mapping of each transition id to info. 
+    mapping(bytes32 => Transitions) transitions;
+
+    //Struct to hold elements of each state
+    struct States {
+        bytes32[] nextStates;
+        mapping(bytes4=>bool) allowedFunctions; 
+    }
```
Comment (alicevhenshaw) [inline contracts/StateMachine.sol]: I think I would keep State as just 2 separate mappings (of stateId to nextStates and stateId to allowedFunctions) as you never actually access the 2 fields in the same situation - they provide very different functionality to one another.

File: contracts/StateMachine.sol [hunk 4]
```diff
@@ -1,141 +1,147 @@
...
+    /// @param _nextStateId stateId of the state to transition to
+    function goToNextState(bytes32 _nextStateId) internal {
+        bytes32 transitionId = getTransitionId(currentStateId, _nextStateId);
+        require(transitions[transitionId].transitionExists);
+        function() internal[] theEffects = transitions[transitionId].transitionEffects;
+        for (uint256 i = 0; i < theEffects.length; i++) {
+            theEffects[i]();
+        }
+        currentStateId = _nextStateId;
+        if (!machineImmutable) {
```
Comment (alicevhenshaw) [inline contracts/StateMachine.sol]: We decided recently we would rather have machine immutability enforced by someone calling a function, not by it automatically happening - so a function called makeImmutable() that makes the state machine immutable from that point onwards, as making it happen automatically causes some issues.

File: contracts/StateMachine.sol [hunk 5]
```diff
@@ -1,141 +1,147 @@
...
+    function allowFunction(bytes32 _stateId, bytes4 _functionSelector) internal stillSettingUp {
+        states[_stateId].allowedFunctions[_functionSelector] = true;
+    }
+
+    /// @dev Goes to the next state if possible (if the next state is valid and reachable by a transition from the current state)
+    /// @param _nextStateId stateId of the state to transition to
+    function goToNextState(bytes32 _nextStateId) internal {
+        bytes32 transitionId = getTransitionId(currentStateId, _nextStateId);
+        require(transitions[transitionId].transitionExists);
+        function() internal[] theEffects = transitions[transitionId].transitionEffects;
```
Comment (alicevhenshaw) [inline contracts/StateMachine.sol]: You can't actually define an array like this strangely - the solidity compiler can't deal with the word function being used inside a function yet.

Comment (vdrg) [reply contracts/StateMachine.sol]: it is supported, the problem is solidity-coverage :cry:

**Comment by vdrg** [comment]
I'm closing this for now until we define how to handle both versions :)

**Pull request closed without merging**